### ISSUE TITLE
wrong language warning (low priority, but still)

### ISSUE THREAD

**Issue by rbolink** [validator/validator#392]
https://validator.w3.org/nu/?doc=http%3A%2F%2Fwww.johanbolink.nl%2F gives the following warning:

Warning: This document appears to be written in Portuguese but the html start tag has lang="nl". Consider using lang="pt" (or variant) instead.
From line 1, column 16; to line 2, column 16
TYPE html>↩<html lang="nl">↩<head

The language is actually Dutch, the domain name is Dutch (.nl) and this is a wrong warning.
There is nothing Portuguese about it.

**Comment by riksoft** [comment]
I've got a similar problem: on pages full of italian text 1 with the word "nostro" (that means "us") makes the validator to issue the warning: "This document appears to be written in Italian. Consider adding lang="it" (or variant) to the html start tag.".

I think it's best to drop this "feature" completely because it shows only misleading warnings: I was initially thinking that such warning was caused by some wrong unvisible bytes in the page because I thought impossibile that only 1 page was recognised as italian.
In other word, I reckon it's best no warning at all than such inconsistent behaviour.

**Comment by sideshowbarker** [comment]
> I've got a similar problem: on pages full of italian text 1 with the word "nostro" (that means "us") makes the validator to issue the warning

If you’re able to share the URL for one of those pages I can examine it and see if I can fix it. I have already made changes to the source that eliminated a number of the case of misidentifications that have been reported.

**Comment by sideshowbarker** [comment]
@rbolink thanks for reporting the problem with https://validator.w3.org/nu/?doc=http%3A%2F%2Fwww.johanbolink.nl%2F and sorry for not responding to this before now. I’m not ignoring it but just have not been able to fix it yet. The language detector unfortunately seems to have difficulty correctly identifying the language in cases like that—documents that are primarily images with a small amount of text for each image. But it is my intention to fix it.

**Comment by sideshowbarker** [comment]
For better or worse, what I’m leaning toward here is: In cases where theTLD is a country code, the checker will never report a warning if the `html lang` value matches what would be expected as the language for that country code. That would at least eliminate the misidentification for the the case of https://validator.w3.org/nu/?doc=http%3A%2F%2Fwww.johanbolink.nl%2F at least. And @rbolink if the documents you mention have a `.it` TLD then it would eliminate those misidentifications too.

**Comment by riksoft** [comment]
Have a look at this page with the warning
http://www.sitoperagenziaimmobiliare.work/
and compare it with this other one longer but without the warning
http://www.sitoperagenziaimmobiliare.work/sitoperagenzieimmobiliari-faq.php

Why only the word "nostro" among all the others is the only one that triggers the error? If I remove that word everything is fine.

**Comment by sideshowbarker** [comment]
> Have a look at this page with the warning
> http://www.sitoperagenziaimmobiliare.work/

I guess I now do not understand what problem you’re reporting. Is that document actually in Italian? If so, then it seems like the language detector is working as expected on that document: It is telling you the page seems to be in Italian but the `html` start tag has no `lang="it"`.

So there you can eliminate the warning by changing line 2 of that document to `<html lang="it">`.

> and compare it with this other one longer but without the warning
> http://www.sitoperagenziaimmobiliare.work/sitoperagenzieimmobiliari-faq.php
> 
> Why only the word "nostro" among all the others is the only one that triggers the error? If I remove that word everything is fine.

Dunno, but I would guess that it might be because adding the word _nostro_ increases the number of characters in the document by just enough that it triggers the language detector to check it.

That is because the language detector will only try to determine the language documents that have at least 1024 characters of non-whitespace text.

**Comment by riksoft** [comment]
> I guess I now do not understand what problem you’re reporting. Is that document actually in Italian? If so, then it seems like the language detector is working as expected on that document: It is telling you the page seems to be in Italian but the html start tag has no lang="it".

The problem is that only 1 page on zillions I've validated give such warning, so I  ask myself: "Why the other million pages have never had this warning? Probably because this one has a hidden hex caracter and the validator get crazy".
I cannot take the warning for what it is because otherwise tons of other pages should have the same warnings.

When cutting and cutting I finally discovered that the culprit was the word "nostro" it confirmed even more that should be some hidden character in there.
That's why I got an hex editor looking around such word. But everything was fine, and at the end I discovered that EVERYWHERE I put the word "nostro" appears the warning.

> Dunno, but I would guess that it might be because adding the word nostro increases the number of characters
I don't think so... check this
http://www.sitoperagenziaimmobiliare.work/sitoperagenzieimmobiliari-faq.php
There is surely more text than in
http://www.sitoperagenziaimmobiliare.work/
but the first is OK and the 2nd get the warning.

Now I know it's only a problem of the validator that give such warning only in particular conditions, but I bet many others will scratch their head when in a bulk validation only 1 page out of 100 get the warning,
